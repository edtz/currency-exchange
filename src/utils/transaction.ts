import { IWallet, ITransaction } from '../feature/Wallet';
import Decimal from 'decimal.js';

type ICreateTransactionReturn = {
    valid: boolean;
    transaction?: ITransaction;
    toBalance?: string;
    fromBalance?: string;
    amounts?: {
        fromAmount: string;
        toAmount: string;
    };
};

let id = 0;
export const createTransaction = (
    fromWallet: IWallet,
    toWallet: IWallet,
    rates: Record<string, number>,
    fromAmount?: number | string,
    toAmount?: number | string,
): ICreateTransactionReturn => {
    const invalid = { valid: false };

    let amounts: { [key: string]: Decimal } = {};
    const fromDecimalBalance = new Decimal(fromWallet.amount);
    const toDecimalBalance = new Decimal(toWallet.amount);
    const decimalRate = new Decimal(rates[toWallet.currency] || 0);

    if ((fromAmount === undefined && toAmount === undefined) || (fromAmount !== undefined && toAmount !== undefined))
        return invalid;

    if (fromAmount === undefined) {
        amounts = {
            fromAmount: new Decimal(toAmount || 0).dividedBy(decimalRate),
            toAmount: new Decimal(toAmount || 0),
        };
    }
    if (toAmount === undefined) {
        amounts = {
            fromAmount: new Decimal(fromAmount || 0),
            toAmount: new Decimal(fromAmount || 0).times(decimalRate),
        };
    }
    if (!amounts.fromAmount.greaterThan(0) || !amounts.toAmount.greaterThan(0)) return invalid;
    const resultingBalances = {
        fromBalance: fromDecimalBalance.minus(amounts.fromAmount),
        toBalance: toDecimalBalance.plus(amounts.toAmount),
    };
    if (resultingBalances.fromBalance.greaterThan(0)) {
        return {
            valid: true,
            transaction: {
                id: id++,
                ts: Date.now(),
                fromWallet: fromWallet.id,
                toWallet: toWallet.id,
                fromAmount: amounts.fromAmount.toFixed(2),
                toAmount: amounts.toAmount.toFixed(2),
            },
            fromBalance: resultingBalances.fromBalance.toFixed(2),
            toBalance: resultingBalances.toBalance.toFixed(2),
            amounts: {
                fromAmount: amounts.fromAmount.toFixed(2),
                toAmount: amounts.toAmount.toFixed(2),
            },
        };
    }
    return invalid;
};
