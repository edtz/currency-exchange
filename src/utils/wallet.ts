import { ITransaction, IWallet } from "../feature/Wallet";

let id = 0;
export const createWallet = (currency: string, amount: string = '0.00', transactionList: ITransaction[] = []): IWallet => ({
    id: id++,
    currency,
    amount,
    transactionList,
});