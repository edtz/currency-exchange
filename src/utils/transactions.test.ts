import { createTransaction } from './transaction';
import { createWallet } from './wallet';

Date.now = jest.fn(() => 1558281600000);

describe('createTransaction():', () => {
    describe('correct data', () => {
        it('should work as expected (1:1)', () => {
            const { transaction, valid } = createTransaction(
                createWallet('EUR', '100'),
                createWallet('GBP'),
                { GBP: 1 },
                10,
            );
            expect(valid).toBe(true);
            expect(transaction).toMatchSnapshot();
        });
        it('should work as expected (1:0.123)', () => {
            const { transaction, valid } = createTransaction(
                createWallet('EUR', '100'),
                createWallet('GBP'),
                { GBP: 0.123 },
                10,
            );
            expect(valid).toBe(true);
            expect(transaction).toMatchSnapshot();
        });
        it('should work as expected (1:99)', () => {
            const { transaction, valid } = createTransaction(
                createWallet('EUR', '100'),
                createWallet('GBP'),
                { GBP: 99 },
                10,
            );
            expect(valid).toBe(true);
            expect(transaction).toMatchSnapshot();
        });
        it('set amount via toAmount', () => {
            const { transaction, valid } = createTransaction(
                createWallet('EUR', '100'),
                createWallet('GBP'),
                { GBP: 1.33 },
                10,
            );
            expect(valid).toBe(true);
            expect(transaction).toMatchSnapshot();
        });
    });
    describe('incorrect data', () => {
        it('empty wallet', () => {
            const { valid } = createTransaction(createWallet('EUR', '0'), createWallet('GBP'), { GBP: 1 }, 10);
            expect(valid).toBe(false);
        });
        it('negative receiver balance', () => {
            const { valid } = createTransaction(createWallet('EUR', '1'), createWallet('GBP', '-1000'), { GBP: 1 }, 1);
            expect(valid).toBe(false);
        });
        it('no amount', () => {
            const { valid } = createTransaction(createWallet('EUR', '1'), createWallet('GBP', '1'), { GBP: 1 });
            expect(valid).toBe(false);
        });
        it('double amount', () => {
            const { valid } = createTransaction(createWallet('EUR', '1'), createWallet('GBP', '1'), { GBP: 1 }, 1, 1);
            expect(valid).toBe(false);
        });
    });
});
