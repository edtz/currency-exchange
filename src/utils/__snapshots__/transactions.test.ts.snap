// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTransaction(): correct data set amount via toAmount 1`] = `
Object {
  "fromAmount": "10",
  "fromWallet": 6,
  "id": 3,
  "toAmount": "13",
  "toWallet": 7,
  "ts": 1558281600000,
}
`;

exports[`createTransaction(): correct data should work as expected (1:0.123) 1`] = `
Object {
  "fromAmount": "10",
  "fromWallet": 2,
  "id": 1,
  "toAmount": "1.2",
  "toWallet": 3,
  "ts": 1558281600000,
}
`;

exports[`createTransaction(): correct data should work as expected (1:1) 1`] = `
Object {
  "fromAmount": "10",
  "fromWallet": 0,
  "id": 0,
  "toAmount": "10",
  "toWallet": 1,
  "ts": 1558281600000,
}
`;

exports[`createTransaction(): correct data should work as expected (1:99) 1`] = `
Object {
  "fromAmount": "10",
  "fromWallet": 4,
  "id": 2,
  "toAmount": "9.9e+2",
  "toWallet": 5,
  "ts": 1558281600000,
}
`;
